modules/rostests/apitests/shell32/LocaleTests.cpp

```diff
@@ -0,0 +1,290 @@
+    { { SH32_PROGRAMS, 1 }, { UENV_STARTUP, 1 } },
+    { { SH32_PROGRAMS, 1 }, { EOLD_PROGRAMS, 0 } },
+    // StartUp
+    { { SH32_STARTUP, 2 }, { UENV_STARTUP, 2 } },
+    // Program Files
+    { { SH32_PROGRAM_FILES, 0 }, { SH32_PROGRAM_FILES_COMMON, 0 } },
+    { { SH32_PROGRAM_FILES, 0 }, { SYSS_PROGRAMFILES, 1 } },
+    // Common Files
+    { { SH32_PROGRAM_FILES_COMMON, 1 }, { SYSS_COMMONFILES, 0 } },
+};
+
+static DWORD CountParts(_In_ LPWSTR str)
+{
+    DWORD count = 0;
+    LPWSTR ptr = str;
+
+    if (*ptr == UNICODE_NULL)
+        return 0;
+
+    while ((ptr = wcschr(ptr, L'\\')) != NULL)
+    {
+        count++;
+        ptr++;
+    }
+
+    return count + 1;
+}
+
+static LPWSTR GetPart(_In_ LPWSTR str, _In_ SIZE_T num, _Out_ SIZE_T* len)
+{
+    DWORD count = 0;
+    LPWSTR ptr = str, next;
+
+    while (count < num && (ptr = wcschr(ptr, L'\\')) != NULL)
+    {
+        count++;
+        ptr++;
+    }
+
+    next = wcschr(ptr, L'\\');
+    *len = next ? next - ptr : wcslen(ptr);
+    return ptr;
+}
+
+static BOOL CALLBACK find_locale_id_callback(
+    _In_ HMODULE hModule, _In_ LPCWSTR type, _In_ LPCWSTR name, _In_ LANGID lang, _In_ LPARAM lParam)
+{
+    langs.insert(lang);
+    return TRUE;
+}
+
+static void SetLocale(_In_ LCID lcid)
+{
+    SetThreadLocale(lcid);
+    SetThreadUILanguage(lcid);
+    curLcid = lcid;
+}
+
+static void TEST_NumParts(void)
+{
+    for (auto i = parts.begin(); i != parts.end(); i++)
+    {
+        E_MODULE m = i->second.eModule;
+
+        if (!mod[m])
+        {
+            skip("No module for test #%d\n", i->first);
+            continue;
+        }
+
+        WCHAR szBuffer[MAX_PATH];
+
+        LoadStringW(mod[m], i->second.id, szBuffer, _countof(szBuffer));
+        ok(i->second.nParts == CountParts(szBuffer), "Locale 0x%lX: Num parts mismatch #%d - expected %lu, got %lu\n",
+           curLcid, i->first, i->second.nParts, CountParts(szBuffer));
+    }
+}
+
+static BOOL LoadPart(_In_ PART* p, _Out_ LPWSTR str, _In_ SIZE_T size)
+{
+    auto s = parts[p->Num];
+    E_MODULE m = s.eModule;
+
+    if (!mod[m])
+        return FALSE;
+
+    WCHAR szBuffer[MAX_PATH];
+    LPWSTR szPart;
+    SIZE_T len;
+
+    LoadStringW(mod[m], s.id, szBuffer, _countof(szBuffer));
+    szPart = GetPart(szBuffer, p->Idx, &len);
+    StringCchCopyNW(str, size, szPart, len);
+
+    return TRUE;
+}
+
+static void TEST_PartMatches(void)
+{
+    for (SIZE_T i = 0; i < _countof(PartMatches); i++)
+    {
+        WCHAR szP1[MAX_PATH], szP2[MAX_PATH];
+
+        if (!LoadPart(&PartMatches[i].p1, szP1, _countof(szP1)))
+        {
+            skip("No module for match test #%lu pair 1\n", i);
+            continue;
+        }
+
+        if (!LoadPart(&PartMatches[i].p2, szP2, _countof(szP2)))
+        {
+            skip("No module for match test #%lu pair 2\n", i);
+            continue;
+        }
+
+        ok(wcscmp(szP1, szP2) == 0, "Locale 0x%lX: Mismatching pairs #%u:%u / #%u:%u '%S' vs. '%S'\n",
```

Comment:
Is it possible to have locale ID in decimal instead of hexadecimal or is this by design?

Reply:
Hexadecimal locale is more recognizable for me 🤔

Reply:
After all I'm going to fix all test failures once the API test is in master.